libyul/backends/evm/SSACFGLiveness.cpp

```diff
@@ -0,0 +1,158 @@
+		std::set<SSACFG::ValueId> live{};
+		block.forEachExit(
+			[&](SSACFG::BlockId const& _successor)
+			{
+				for (auto const& phi: m_cfg.block(_successor).phis)
+				{
+					auto const& info = m_cfg.valueInfo(phi);
+					yulAssert(std::holds_alternative<SSACFG::PhiValue>(info), "value info of phi wasn't PhiValue");
+					auto const& entries = m_cfg.block(std::get<SSACFG::PhiValue>(info).block).entries;
+					// this is getting the argument index of the phi function corresponding to the path going
+					// through "blockId", ie, the currently handled block
+					auto const it = entries.find(blockId);
+					yulAssert(it != entries.end());
+					auto const argIndex = static_cast<size_t>(std::distance(entries.begin(), it));
+					if (argIndex < std::get<SSACFG::PhiValue>(info).arguments.size())
+					{
+						// todo not sure why arg index would exceed phi arg length but it does happen
+						auto const arg = std::get<SSACFG::PhiValue>(info).arguments.at(argIndex);
+						if (!std::holds_alternative<SSACFG::LiteralValue>(m_cfg.valueInfo(arg)))
+							live.insert(arg);
+					}
+					else
+						yulAssert(false);
+				}
+			});
+
+		// for each S \in succs(B) s.t. (B, S) not a back edge: live <- live \cup (LiveIn(S) - PhiDefs(S))
+		block.forEachExit(
+			[&](SSACFG::BlockId const& _successor) {
+				if (!m_topologicalSort.backEdge(blockId, _successor))
+					live += m_liveIns[_successor.value] - m_cfg.block(_successor).phis;
+			});
+		if (std::holds_alternative<SSACFG::BasicBlock::FunctionReturn>(block.exit))
+			live += std::get<SSACFG::BasicBlock::FunctionReturn>(block.exit).returnValues
+					| ranges::views::filter(filterLiterals);
+
+		// clean out unreachables
+		live = live | ranges::views::filter([&](auto const& valueId) { return !std::holds_alternative<SSACFG::UnreachableValue>(m_cfg.valueInfo(valueId)); }) | ranges::to<std::set>;
+
+		// LiveOut(B) <- live
+		m_liveOuts[blockId.value] = live;
+
+		// for each program point p in B, backwards, do:
+		for (auto const& op: block.operations | ranges::views::reverse)
+		{
+			// remove variables defined at p from live
+			live -= op.outputs | ranges::views::filter(filterLiterals) | ranges::to<std::vector>;
+			// add uses at p to live
+			live += op.inputs | ranges::views::filter(filterLiterals) | ranges::to<std::vector>;
+		}
+
+		// livein(b) <- live \cup PhiDefs(B)
+		m_liveIns[blockId.value] = live + block.phis;
+	}
+}
+
+void SSACFGLiveness::runLoopTreeDfs(size_t const _loopHeader)
+{
+	// SSA Book, Algorithm 9.3
+	if (m_loopNestingForest.loopNodes().count(_loopHeader) > 0)
+	{
+		// the loop header block id
+		auto const& block = m_cfg.block(SSACFG::BlockId{_loopHeader});
+		// LiveLoop <- LiveIn(B_N) - PhiDefs(B_N)
+		auto liveLoop = m_liveIns[_loopHeader] - block.phis;
+		// must be live out of header if live in of children
+		m_liveOuts[_loopHeader] += liveLoop;
+		// for each blockId \in children(loopHeader)
+		for (size_t blockId = 0; blockId < m_cfg.numBlocks(); ++blockId)
```

Comment:
Since you used `blockIdValue` in other places, it might make sense to maintain that distinction here as well.
```suggestion
		for (size_t blockIdValue = 0; blockIdValue < m_cfg.numBlocks(); ++blockIdValue)
```